File: AbstractFactory/AbstractFactory.java
interface Car {
    void say();
}

abstract class AudiCar implements Car {
}

class AudiA4 extends AudiCar {
    public void say() {
        System.out.println("I am Audi A4");
    }
}

class AudiQ5 extends AudiCar {
    public void say() {
        System.out.println("I am Audi Q5");
    }
}

abstract class BMWCar implements Car {
}

class BMWX1 extends BMWCar {
    public void say() {
        System.out.println("I am BMW X1");
    }
}

class BMW3X extends BMWCar {
    public void say() {
        System.out.println("I am BMW 3X");
    }
}

abstract class CarFactory {
    public AudiCar createAudi() {
        return createAudiCar();
    }
    public AudiCar createAudiCar() {
        return null;
    }

    public BMWCar createBMW() {
        return createBMWCar();
    }
    public BMWCar createBMWCar() {
        return null;
    }
}

class AudiA4Factory extends CarFactory {
    public AudiCar createAudiCar() {
        return new AudiA4();
    }
}

class AudiQ5Factory extends CarFactory {
    public AudiCar createAudiCar() {
        return new AudiQ5();
    }
}

class BMWX1Factory extends CarFactory {
    public BMWCar createBMWCar() {
        return new BMWX1();
    }
}

class BMW3XFactory extends CarFactory {
    public BMWCar createBMWCar() {
        return new BMW3X();
    }
}


public class AbstractFactory {
    public static void main(String[] args) {
        AudiCar a4 = new AudiA4Factory().createAudi();
        AudiCar q5 = new AudiQ5Factory().createAudi();
        BMWCar x1 = new BMWX1Factory().createBMW();
        BMWCar _3x = new BMW3XFactory().createBMW();

        a4.say();
        q5.say();
        x1.say();
        _3x.say();
    }
}

File: AbstractFactory/AbstractFactory.php
<?php
/**
 * Created by PhpStorm.
 * User: lanjialiang
 * Date: 2017/8/23
 * Time: 下午4:02
 */


interface CarInterface{
    public function desc();
}
abstract class AudiCar implements CarInterface{}
abstract class BMWCar implements CarInterface{}

class AudiA4 extends AudiCar{
    const AUDI_A4 = "this is AudiA4";

    public function desc(){
        return self::AUDI_A4;
    }
}


class BMWx1 extends BMwCar{
    const BMW_X1 = "this is BMWx1";

    public function desc(){
        return self::BMW_X1;
    }
}
/**
 * car factory
 */
abstract class CarFactory{
    public  function createAudi(){
        return $this->createAudiCar();
    }

    public function createAudiCar(){
    }

    public function createBMW(){
        return $this->createBMWCar();
    }

    public function createBMWCar(){
    }
}
class AudiA4Factory extends CarFactory{
    public function createAudiCar(){
        return new AudiA4();
    }
}
class BMWx1Factory extends  CarFactory{
    public function createBMWCar(){
        return new BMWx1();
    }
}

class Test{
    public function __invoke(){
        $audiA4 = new AudiA4Factory();
        $bmwX1 = new BMWx1Factory();
        assert(AudiA4::AUDI_A4 === $audiA4->createAudi()->desc(), "audiA4 fail");
        assert(BMWx1::BMW_X1 === $bmwX1->createBMW()->desc(),"bmwX1 fail");
    }
}
$test = new Test();
$test();
